File: app/src/main/java/com/example/dcmtkdemo/view/PacsConnectionView.java
package com.example.dcmtkdemo.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.Nullable;

import com.example.dcmtk.jni.DcmtkJni;
import com.example.dcmtkdemo.databinding.ViewPacsConnectionBinding;

public class PacsConnectionView extends LinearLayout {

    private ViewPacsConnectionBinding binding;
    private OnConnectionVerifiedListener listener;

    public interface OnConnectionVerifiedListener {
        void onVerified(String host, int port, String local, String remote);
    }

    public PacsConnectionView(Context context) {
        this(context, null);
    }

    public PacsConnectionView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public PacsConnectionView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setOrientation(VERTICAL);
        binding = ViewPacsConnectionBinding.inflate(LayoutInflater.from(context), this);
        init();
    }

    private void init() {
        binding.btnTestConnect.setOnClickListener(v -> {
            String host = getHost();
            int port = getPort();
            String local = getLocalAet();
            String remote = getRemoteAet();

            new Thread(() -> {
                boolean success = DcmtkJni.connectPACS(host, port, local, remote);
                post(() -> Toast.makeText(getContext(), "Connection: "
                        + (success ? "OK" : "Failed"), Toast.LENGTH_SHORT).show());
            }).start();
        });

        binding.btnCecho.setOnClickListener(v -> {
            String host = getHost();
            int port = getPort();
            String local = getLocalAet();
            String remote = getRemoteAet();

            new Thread(() -> {
                boolean success = DcmtkJni.cEcho(host, port, local, remote);
                post(() -> {
                    Toast.makeText(getContext(), "C-ECHO: "
                            + (success ? "Success" : "Failed"), Toast.LENGTH_SHORT).show();
                    if (success && listener != null) {
                        listener.onVerified(host, port, local, remote);
                    }
                });
            }).start();
        });
    }

    public void setOnConnectionVerifiedListener(OnConnectionVerifiedListener listener) {
        this.listener = listener;
    }

    public void setConnectionInfo(String host, int port, String local, String remote) {
        binding.etHost.setText(host);
        binding.etPort.setText(String.valueOf(port));
        binding.etLocalAet.setText(local);
        binding.etRemoteAet.setText(remote);
    }

    public String getHost() { return binding.etHost.getText().toString(); }
    public int getPort() {
        try {
            return Integer.parseInt(binding.etPort.getText().toString());
        } catch (Exception e) {
            return 0;
        }
    }
    public String getLocalAet() { return binding.etLocalAet.getText().toString(); }
    public String getRemoteAet() { return binding.etRemoteAet.getText().toString(); }
}

File: dcmtk/src/main/java/com/example/dcmtk/view/PacsConnectionView.kt
package com.example.dcmtk.view

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import com.example.dcmtk.PacsManager
import com.example.dcmtk.databinding.ViewPacsConnectionBinding
import com.example.dcmtk.model.PacsConfig
import kotlinx.coroutines.*

class PacsConnectionView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    private val binding: ViewPacsConnectionBinding =
        ViewPacsConnectionBinding.inflate(LayoutInflater.from(context), this)
    private var listener: OnConnectionVerifiedListener? = null
    
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    interface OnConnectionVerifiedListener {
        fun onVerified(config: PacsConfig)
        fun onCancel()
    }

    init {
        orientation = VERTICAL
        initView()
    }

    private fun initView() {
        binding.btnVerify.setOnClickListener {
            val config = getConfig() ?: return@setOnClickListener

            binding.tvError.visibility = View.GONE
            binding.btnVerify.isEnabled = false
            binding.btnCancel.isEnabled = false

            scope.launch {
                val echoSuccess = PacsManager.safeCEcho(config, 2)
                
                if (echoSuccess) {
                    binding.tvError.visibility = View.GONE
                    Toast.makeText(context, "Connection & C-ECHO Verified", Toast.LENGTH_SHORT).show()
                    listener?.onVerified(config)
                } else {
                    showError("Verification Failed (Check Network or AETs)")
                }
                binding.btnVerify.isEnabled = true
                binding.btnCancel.isEnabled = true
            }
        }

        binding.btnCancel.setOnClickListener {
            listener?.onCancel()
        }
    }

    fun setOnConnectionVerifiedListener(listener: OnConnectionVerifiedListener?) {
        this.listener = listener
    }

    private fun showError(message: String) {
        binding.tvError.text = message
        binding.tvError.visibility = View.VISIBLE
    }

    fun setConnectionInfo(config: PacsConfig?) {
        binding.etHost.setText(config?.host ?: "")
        binding.etPort.setText(config?.port?.toString() ?: "11112")
        binding.etLocalAet.setText(config?.localAet ?: "")
        binding.etRemoteAet.setText(config?.remoteAet ?: "")
    }

    fun setConnectionInfo(host: String?, port: Int, local: String?, remote: String?) {
        setConnectionInfo(PacsConfig(host ?: "", port, local ?: "", remote ?: ""))
    }

    fun getConfig(): PacsConfig? {
        val host = binding.etHost.text.toString().trim()
        val portStr = binding.etPort.text.toString().trim()
        val local = binding.etLocalAet.text.toString().trim()
        val remote = binding.etRemoteAet.text.toString().trim()

        if (host.isEmpty() || portStr.isEmpty() || local.isEmpty() || remote.isEmpty()) {
            showError("Please fill all fields")
            return null
        }

        val port = portStr.toIntOrNull() ?: 0
        if (port <= 0) {
            showError("Invalid port number")
            return null
        }

        return PacsConfig(host, port, local, remote)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        scope.cancel()
    }
}
